pennylane/beta/plugins/default_tensor.py

```diff
@@ -572,73 +636,82 @@ def ev(self, obs_nodes, wires):
-        for w in set(all_wires) - set(meas_wires):
-            self._add_edge(bra, w, ket, w)  # |psi[w]|**2
-
-        # At this stage, all nodes are connected, and the contraction yields a
-        # scalar value.
-        contracted_ket = ket
-        for obs_node in obs_nodes:
-            contracted_ket = tn.contract_between(obs_node, contracted_ket)
-        expval = tn.contract_between(bra, contracted_ket).tensor
-        if self._abs(self._imag(expval)) > tolerance:
+        if self._rep == "exact":
+            self._contract_to_ket(method="auto")
+            ket = self._nodes["contracted_state"]
+            bra = tn.conj(ket, name="Bra")
+
+            all_wires = tuple(w for w in range(self.num_wires))
+            meas_wires = []
+            # We need to build up <psi|A|psi> step-by-step.
+            # For wires which are measured, we need to connect edges between
+            # bra, obs_node, and ket.
+            # For wires which are not measured, we need to connect edges between
+            # bra and ket.
+            # We use the convention that the indices of a tensor are ordered like
+            # [output_idx1, output_idx2, ..., input_idx1, input_idx2, ...]
+            for obs_node, obs_wires in zip(obs_nodes, wires):
+                meas_wires.extend(obs_wires)
+                for idx, w in enumerate(obs_wires):
+                    output_idx = idx
+                    input_idx = len(obs_wires) + idx
+                    self._add_edge(obs_node, input_idx, ket, w)  # A|psi>
+                    self._add_edge(bra, w, obs_node, output_idx)  # <psi|A
+            for w in set(all_wires) - set(meas_wires):
+                self._add_edge(bra, w, ket, w)  # |psi[w]|**2
+
+            # At this stage, all nodes are connected, and the contraction yields a
+            # scalar value.
+            contracted_ket = ket
+            for obs_node in obs_nodes:
+                contracted_ket = tn.contract_between(obs_node, contracted_ket)
+            expval = tn.contract_between(bra, contracted_ket).tensor
+
+        elif self._rep == "mps":
+            raise NotImplementedError
+            # if len(wires) == 1:
+            #    expval = self.mps.measure_local_operator(obs_nodes, wires[0])
+            # else:
+            #    raise NotImplementedError
+
+        if self._abs(self._imag(expval)) > TOL:
             warnings.warn(
                 "Nonvanishing imaginary part {} in expectation value.".format(expval.imag),
                 RuntimeWarning,
             )
         return self._real(expval)
 
+    def _contract_to_ket(self, method="auto"):
+        """Contract the nodes which represent the state preparation and evolution to a final tensor.
+        The contracted tensor is stored in the `_nodes` dictionary under the key `"contracted_state"`
+
+        Args:
+            method (str): the contraction method to be employed (possible choices are listed
+             in the dictionary `contract_fns`)
```

Comment:
```suggestion
            method (str): the contraction method to be employed (possible choices are listed
                in the dictionary `contract_fns`)
```

Reply:
Note: the `contract_fns` dictionary likely won't appear in the documentation as it's a module level variable. Better to describe the options here in this docstring.